apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -49,3 +39,225 @@
 - Missing $this
 
 */
+
+use apollo_compiler::schema::{Component, Directive};
+use apollo_compiler::Schema;
+use itertools::Itertools;
+use std::fmt::Display;
+use url::Url;
+
+/// Validate the connectors-related directives `@source` and `@connect`.
+///
+/// This function attempts to collect as many validation errors as possible, so it does not bail
+/// out as soon as it encounters one.
+pub fn validate(schema: Schema) -> Vec<ValidationError> {
+    let mut source_names = Vec::new();
+    let errors = schema
+        .schema_definition
+        .directives
+        .iter()
+        .filter_map(|directive| {
+            if directive.name == "source" {
+                match validate_source(directive) {
+                    Ok(source_directive) => {
+                        if let SourceName::Valid(name) = source_directive.name {
+                            source_names.push(name);
```

Comment:
Nit: I would avoid having functional components with side-effects, if possible. You could try using a fold or something and having the map return a tuple, if you prefer.

Reply:
This will inevitably change when it becomes more complex once connect directives are also added—since there's going to be cross-validation. So I'll wait to clean up this code until we see what it needs.